networkx/algorithms/bipartite/matching.py

```diff
@@ -582,3 +585,310 @@ def minimum_weight_full_matching(G, top_nodes=None, weight="weight"):
     # add the ones from right to left as well.
     d.update({v: u for u, v in d.items()})
     return d
+
+
+def _neighbours_of_set(G, node_set):
+    r"""Returns a set of the neighbours of a given set of nodes `node_set` in the bipartite graph `G`.
+    Denoted by :math:`N_Gnode_set`.
+
+    `G` is an undirected bipartite NetworkX graph.
+
+    `node_set` is a set of nodes in `G`.
+
+
+    >>> G = nx.complete_bipartite_graph(3,3)
+    >>> _neighbours_of_set(G, {})
+    set()
+    >>> _neighbours_of_set(G, {1, 2})
+    {3, 4, 5}
+
+    >>> G = nx.Graph([(0, 4), (1, 5), (2, 6)])
+    >>> _neighbours_of_set(G, {0, 1})
+    {4, 5}
+
+    >>> G=nx.Graph([(0,3),(3,0),(0,4),(4,0),(1,4),(4,1),(2,4),(4,2)])
+    >>> _neighbours_of_set(G, {0, 1})
+    {3, 4}
+
+    >>> _neighbours_of_set(G, {4})
+    {0, 1, 2}
+    """
+    ret_set = {}
+    for node in node_set:
+        ret_set.update(G[node])
+
+    return set(ret_set)
+
+
+def _M_alternating_sequence(G, M, *, top_nodes=None):
+    r"""Generates a M-alternating-sequence of a bipartite graph `G` with regard to a matching M.
+
+    Two sets are calculated with the following recursive definition:
+
+    X_0 = nodes in G unmatched by M.
+    for i>=1:
+        Latex:
+            Y_i = N_{G\M}(X_{i-1})
+            X_i = N_M(Y_i)
+        English:
+            Y_i -> neighbours of X_{i-1} that were not matched by M group-subtracting all the previous Y_k's we've seen
+            X_i -> neighbours of Y_i that were matched by M
+
+    The iterative pattern stops when either Y_k or X_k are an empty set for a certain k, forever staying an empty set from k onward
+    the alternating sequece is proven to always stop in the article.
+
+    >>> G = nx.complete_bipartite_graph(3,3)
+    >>> M = nx.bipartite.hopcroft_karp_matching(G)
+    >>> _M_alternating_sequence(G, M)
+    ((), ())
+
+    >>> G = nx.Graph([(0,3),(3,0),(0,4),(4,0),(1,4),(4,1),(2,4),(4,2)])
+    >>> M = {0:3, 3:0, 1:4, 4:1}
```

Comment:
Why exactly is the example bit written like this? For an undirected graph, adding an edge `0 -> 3` will also add `3 -> 0` so something like `nx.Graph([(0,3),(0,4),(1,4),(2,4)])` is equivalent to the above example.

Reply:
Correct. I removed these redundancies in all places where I noticed them in the docstrings.
I'm guessing the matching should stay as it is, I noticed the existing networkx matching algorithms do return a matching of this form.